pim, pimv6: Removing upstreams and mroutes when pim configs removed on receiver connected interface.

SaiGomathiN opened the pull request:
When the (no ip pim/no ipv6 pim) pim/pimv6 configuration is removed from the receiver-connected interface in LHR, we observe inconsistent behavior in following scenarios.

1) When there is no pim/pimv6 configuration at the beginning and the traffic is initiated, no multicast routes are formed, and there is no available upstream 

```
frr# show ip pim upstream
 Iif  Source  Group  State  Uptime  JoinTimer  RSTimer  KATimer  RefCnt

frr# show ip mroute
IP Multicast Routing Table
Flags: S - Sparse, C - Connected, P - Pruned
       R - SGRpt Pruned, F - Register flag, T - SPT-bit set
 Source  Group  Flags  Proto  Input  Output  TTL  Uptime
```

2) When the traffic is flowing and we remove the pim/pimv6 config in between, the (*,G) upstream transitions into the NotJ state, yet the mroutes remain intact.

```
frr# show ip mroute
IP Multicast Routing Table
Flags: S - Sparse, C - Connected, P - Pruned
       R - SGRpt Pruned, F - Register flag, T - SPT-bit set
 Source  Group      Flags  Proto  Input   Output  TTL  Uptime
 *       226.1.1.1  SC     IGMP   ens224  pimreg  1    00:00:56
                           IGMP           ens192  1

frr# show ip pim upstream
 Iif     Source  Group      State  Uptime    JoinTimer  RSTimer   KATimer   RefCnt
 ens224  *       226.1.1.1  NotJ   00:00:13  00:00:55   --:--:--  --:--:--  1

```
Hence, to make the behaviour consistent, we remove both the upstream entries for (*,G) and the corresponding mroutes using the 'pim_igmp_if_reset' API call when dealing with the receiver-connected interface. The 'pim_ifchannel_local_membership_del' is returned when no ip pim/pimv6 config is applied due to 'pim_ifp->pim_enable' check. Therefore, removing this check in order to proceed with the deletion when the pim/pimv6 configurations are removed.

Closes #13890

SaiGomathiN committed:
pim: Modifying pim_igmp_if_reset api name to common name

Modifying pim_igmp_if_reset to pim_gm_if_reset as common name
and paramter to struct interface for accomodating both IGMP and MLD code changes.

Signed-off-by: Sai Gomathi N <nsaigomathi@vmware.com>

ton31337 reviewed (commented):
Just a qq: isn't this relevant for MLD also?

SaiGomathiN commented:
ci:rerun

SaiGomathiN commented:
ci:rerun ospf failures not related to this change.

SaiGomathiN commented:
ci:rerun some tests are incomplete

SaiGomathiN commented:
ci:rerun bgp related failures

SaiGomathiN commented:
ci:rerun unrelated failures

SaiGomathiN commented:
ci:rerun

SaiGomathiN commented:
ci:rerun

mobash-rasool commented on pimd/pim_igmp.h:
```diff
@@ -127,6 +127,10 @@ static inline void pim_igmp_other_querier_timer_on(struct gm_sock *igmp)
 {
 }
+
+static inline void pim_igmp_if_reset(struct pim_interface *pim_ifp)
```
In case of pimv6 here, call gm_group_delete function inside here. So we can remove the if else check in the caller. Change the api name to pim_gm_if_reset as well.

mobash-rasool commented on pimd/pim_pim.c:
```diff
@@ -124,6 +131,15 @@ void pim_sock_delete(struct interface *ifp, const char *delete_message)
 	sock_close(ifp);
+
+	if (pim_ifp->gm_enable) {
+#if PIM_IPV == 4
```
we can remove if else checks here by calling pim_gm_if_reset directly as mentioned above.

mobash-rasool commented:
> Just a qq: isn't this relevant for MLD also?

Yes @ton31337 : That is the reason you can see #ifdef PIM_IPV6 being used.

patrasar commented on pimd/pim_pim.c:
```diff
@@ -124,6 +127,9 @@ void pim_sock_delete(struct interface *ifp, const char *delete_message)
 	pim_neighbor_delete_all(ifp, delete_message);
 
 	sock_close(ifp);
+
```
It would be really good if the below code is moved to Line no. 117.
IMO, we should clean off the group information in our side first then send hello with 0 hold time and clean up the neighbor information.

SaiGomathiN committed:
pimd: Modifying the parameter for pim_igmp_if_fini api

Modifying the parameter as struct interface *ifp to accomodate
the MLD and IGMP changes for pim_igmp_if_reset api

Signed-off-by: Sai Gomathi N <nsaigomathi@vmware.com>

SaiGomathiN committed:
pim, pimv6: Removing upstreams and mroutes when "no ip pim"/"no ipv6 pim" configs applied in receiver connected interface

When the (no ip pim/no ipv6 pim) pim/pimv6 configuration is removed from the receiver-connected interface in LHR,
we observe inconsistent behavior in following scenarios.

1) When there is no pim/pimv6 configuration at the beginning and the traffic is initiated,
no multicast routes are formed, and there is no available upstream
frr# show ip pim upstream
 Iif  Source  Group  State  Uptime  JoinTimer  RSTimer  KATimer  RefCnt

frr# show ip mroute
IP Multicast Routing Table
Flags: S - Sparse, C - Connected, P - Pruned
       R - SGRpt Pruned, F - Register flag, T - SPT-bit set
 Source  Group  Flags  Proto  Input  Output  TTL  Uptime

2) When the traffic is flowing and we remove the pim/pimv6 config in between,
the (*,G) upstream transitions into the NotJ state, yet the mroutes remain intact.

frr# show ip mroute
IP Multicast Routing Table
Flags: S - Sparse, C - Connected, P - Pruned
       R - SGRpt Pruned, F - Register flag, T - SPT-bit set
 Source  Group      Flags  Proto  Input   Output  TTL  Uptime
 *       226.1.1.1  SC     IGMP   ens224  pimreg  1    00:00:56
                           IGMP           ens192  1

frr# show ip pim upstream
 Iif     Source  Group      State  Uptime    JoinTimer  RSTimer   KATimer   RefCnt
 ens224  *       226.1.1.1  NotJ   00:00:13  00:00:55   --:--:--  --:--:--  1

Hence, to make the behaviour consistent, we remove both the upstream entries for (*,G)
and the corresponding mroutes using the 'pim_igmp_if_reset' API call when dealing with the receiver-connected interface.
The 'pim_ifchannel_local_membership_del' is returned when no ip pim/pimv6 config is applied due to 'pim_ifp->pim_enable' check.
Therefore, removing this check in order to proceed with the deletion when the pim/pimv6 configurations are removed

Signed-off-by: Sai Gomathi N <nsaigomathi@vmware.com>

Jafaral commented:
update

riw777 commented:
@frrbot autoclose in 2 weeks
